### ISSUE TITLE
About the animation player

### ISSUE THREAD

**Issue by Freedomax** [castle-engine/castle-engine#473]
I have written an [animation player](https://github.com/Freedomax/castle-engine/tree/animation-player), about 5000 lines of code. It has been tested and runs well, serialization and designer have also been completed. 
However, there is an issue:
Animations need to call `Update` every frame, and `TAnimationPlayer` is inherited from `TCastleComponent`. I had to register two additional controls: `TCastleAnimationPlayer = class(TCastleUserInterface)` and `TCastleAnimationPlayerTransform = class(TCastleTransform)`, both of which have TAnimationPlayer as a subcomponent.

I wonder if `TAnimationPlayer` can be registered directly as a `TCastleBehavior` without the need for two additional controls. However, `TCastleUserInterface` cannot use `TCastleBehavior`. Can it be added?

**Comment by michaliskambi** [comment]
Hm, admittedly it's something I thought about in the past and didn't have a good solution. (But I think I have one now, see below :) )

Thoughts:

- A consequence of our current `TCastleBehavior` API is that it really can be attached only to `TCastleTransform` -- it has various methods and fields dealing with `Parent: TCastleTransform`, `World: TCastleAbstractRootTransform`. And they are useful, e.g. `TCastleBillboard` adjusts Parent rotation, `TCastleSoundSource` follows parent position. 

- At the same time, our `TCastleTransform` is something unrelated to `TCastleUserInterface` (unlike e.g. Unity where everything is a GameObject). There are advantages to it (multiple viewports are much more natural in CGE, viewport is a UI that contains transforms, you can have many viewports that share or not share transforms) and disadvantages to it -- this is one disadvantage, that concept like "behavior", while it conceptually makes sense on UI too, is in fact tied to `TCastleTransform`.

I think I have a solution that I would consider clean:

- Introduce `TCastleBaseBehavior`. This would descend from `TCastleComponent` and basically just have an `procedure Update(const SecondsPassed: Single; var RemoveMe: TRemoveType); virtual;` method. Also it would have `Parent: TCastleComponent`.

- Existing `TCastleBehavior` would descend from `TCastleBaseBehavior`. So `TCastleBehavior` stays backward-compatible, and has all the API about parent `TCastleTransform`. It has `Parent: TCastleTransform` that just hides ancestor `Parent: TCastleComponent` -- no problem with that, as `TCastleTransform` is still `Parent: TCastleComponent`, so e.g. doing

    ```
    var
      B: TCastleBehavior;
      BB: TCastleBaseBehavior;
    begin
      B := ...
      WritelnLog(B.Parent.Name);
      BB := B;
      WritelnLog(BB.Parent.Name);
    end;
    ```

   ... reliably always writes the same name. Both properties `TCastleBaseBehavior.Parent` and `TCastleBehavior.Parent` return the same thing, just the latter is type-casted to `TCastleTransform`. 

    Trying to add `TCastleBehavior` to a parent that isn't `TCastleTransform` would raise an exception (and in editor, we can just make it impossible using UI).

- Users can inherit from `TCastleBehavior` if their behavior is naturally connected only to `TCastleTransform`. E.g. `TCastleBillboard`, `TCastleSoundSource` -- they remain `TCastleBehavior` descendants, and it makes sense: there's no point attaching them e.g. to UI.

- Users can inherit from `TCastleBaseBehavior` if all they need is an `Update` method and/or `Parent` as `TCastleComponent`. 

- `TCastleTransform.AddBehavior` and friends have to be extended to allow any `TCastleBaseBehavior`. They do some special stuff when behavior is `TCastleBehavior` (to set world), but they accept any `TCastleBaseBehavior`.

- We add to `TCastleUserInterface` routines to add/remove `TCastleBaseBehavior`.

Let me think about it a bit, but I think I like this approach. 

And then `TAnimationPlayer` can be attached to anything, transform or UI.

**Comment by Freedomax** [comment]
Yes, this is a good solution. Because both UIControl and Transform have an `Update` function, for example, `TCastleTimer` can be inherited from `TCastleBaseBehavior` in the future, and look forwarding to your `TCastleBaseBehavior`.

Additionally, there may be another issue unrelated to this issue that I would like to discuss here. 

Sometimes, I want to inherit some classes from CGE, but I encounter some problems. For example, I wrote a custom control: `TCastleSprite`, which inherits from `TCastleImageTransform` and is similar to the `TSprite` in Godot. It has three integer published properties: `Row`, `Column`, and `Frame`, and updates `Shift`, `RepeatImage`, and `Pivot` internally. With the animation player, this class can be used as a "real sprite", which is very convenient. I only need to update the `Frame` property without manually importing images in the sprite editor.

However, I want to hide `Shift`, `RepeatImage`, and `Pivot`, but currently, it is not possible. There are many **TxxxPersistent** properties for the editor that are actually unnecessary. Therefore, similar to LCL/VCL, `TCastleImageTransform` should have a base class `TCustomImageTransform` for custom control inheritance.

Similarly, the same problem occurs in `TCastleTransform`, and I need a base class `TCustomTransform` for inheritance, which does not need published properties like **TxxxPersistent**. So, is it possible to make these modifications to the engine?

**Comment by michaliskambi** [comment]
> Sometimes, I want to inherit some classes from CGE, but I encounter some problems. For example, I wrote a custom control: TCastleSprite, which inherits from TCastleImageTransform and is similar to the TSprite in Godot. It has three integer published properties: Row, Column, and Frame, and updates Shift, RepeatImage, and Pivot internally. With the animation player, this class can be used as a "real sprite", which is very convenient. I only need to update the Frame property without manually importing images in the sprite editor.
> 
> However, I want to hide Shift, RepeatImage, and Pivot, but currently, it is not possible. There are many TxxxPersistent properties for the editor that are actually unnecessary. Therefore, similar to LCL/VCL, TCastleImageTransform should have a base class TCustomImageTransform for custom control inheritance.
> 
> Similarly, the same problem occurs in TCastleTransform, and I need a base class TCustomTransform for inheritance, which does not need published properties like TxxxPersistent. So, is it possible to make these modifications to the engine?

Here I deliberately decided we'll not create such "custom" classes. Let me explain why and show alternatives :)

Why?

I perceived them as a burden -- it's extra maintenance work, and at the same time it doesn't solve the problem ("you exposed access to something you should not") in 100% clean way.

- One effectively needs to create 2 classes instead of 1 each time. 

- There's no perfect approach to inheritance: You either make all `TCustomXxx` descend from other `TCustomXxx` (like `TCustomSprite` would descend from `TCustomTransform`) and then the final class may have a lot of properties to make published in a standard case. Or `TCustomXxx` descends from non-custom, and then ancestor fields are already public anyway. In either case, the final non-custom clas `TXxx` has some work to do when ancestors get extended -- in the usual case, if the inheritance really makes sense, it wants to publish most of ancestors' stuff.

- It doesn't solve the fundamental issue: class A descends from B, so you have all the stuff from B in A. Not making the unnecessary things published only hides them in editor -- they are still accessible from code (and will be confusing for programmer).

    For example in your case, the programmer would be able to access `Pivot` anyway, as it would be public. Or protected.

    In other example, the `TCustomTransform` would still be actually a transformation, with possible translation/rotation/scale. Editor would want to show pivot for them, at least read-only. 

- I haven't seen non-Pascal frameworks using this approach. E.g. I haven't seen this approach in Unity and C#.

There are better ways:

- If you really only want to hide them, you can override `PropertySections` for CGE editor, to make them disappear from all tabs except "All". So e.g. ancestor may place "Pivot" in "Basic", but descendant can move it away from "Basic".

    You cannot remove the property from "All" tab in CGE editor (maybe it should be called "Advanced"), but you control all other tabs.

- If the properties should not be used, you should hide them in a proper way, which means not using inheritance. Use composition instead. Instead of inheriting A from B, make B a visual child of A. In your case, it would mean that `TCastleSprite` does not inherit from `TCastleImageTransform`, instead your `TCastleSprite` creates `TCastleImageTransform` as a child (with internal reference). And then you can expose the properties you want on `TCastleSprite`.

    This is the approach used many times in CGE. E.g. `TCastleAbstractLight` actually has an internal `TCastleScene` with X3D node of the light. You don't see this `TCastleScene` instance, you don't see this X3D light (though actually you could access it by accessing `MyLight.List[0] as TCastleScene` but then you know you're accessing something that should be internal). Similarly e.g. `TCastleTerrain`, `TCastleAbstractPrimitive` have also a `TCastleScene` child.

    Similarly for UI e.g. `TCastleCheckbox` contains `TCastleLabel` and some images.

**Comment by michaliskambi** [comment]
I definitely will want to add `TCastleBaseBehavior` soon, and make the behavior changes mentioned above. I see more and more benefits of it :)

First the prerequisite is something I planned already, a proper `TCastleComponentList<T>` honored by CGE editor to display the hierarchy -- having it would make such changes much easier. It will avoid having the need to essentially repeat the same/extremely similar logic across a few parent-child relationships in CGE now.
